File: cast-extractor/src/main/java/com/libreshockwave/tools/score/ScoreDataBuilder.java
package com.libreshockwave.tools.score;

import com.libreshockwave.DirectorFile;
import com.libreshockwave.chunks.*;
import com.libreshockwave.tools.format.ChannelNames;
import com.libreshockwave.tools.format.PaletteDescriptions;
import com.libreshockwave.tools.model.ScoreCellData;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Builds score grid data for display.
 */
public class ScoreDataBuilder {

    /**
     * Builds a 2D array of score cell data for a Director file.
     * Rows = channels, Columns = frames.
     */
    public ScoreCellData[][] buildScoreData(DirectorFile dirFile) {
        ScoreChunk scoreChunk = dirFile.getScoreChunk();
        if (scoreChunk == null) {
            return new ScoreCellData[0][0];
        }

        int frameCount = scoreChunk.getFrameCount();
        int channelCount = scoreChunk.getChannelCount();

        ScoreCellData[][] data = new ScoreCellData[channelCount][frameCount];

        // Build a map of frame intervals (scripts attached to frames)
        Map<Integer, String> frameScriptMap = buildFrameScriptMap(dirFile, scoreChunk, frameCount);

        // Populate with frame channel data
        ScoreChunk.ScoreFrameData frameData = scoreChunk.frameData();
        if (frameData != null && frameData.frameChannelData() != null) {
            for (ScoreChunk.FrameChannelEntry entry : frameData.frameChannelData()) {
                int fr = entry.frameIndex();
                int ch = entry.channelIndex();
                ScoreChunk.ChannelData chData = entry.data();

                if (fr >= 0 && fr < frameCount && ch >= 0 && ch < channelCount && !chData.isEmpty()) {
                    // Resolve display name based on channel type
                    String displayName;

                    // For Tempo channel, prefer the frame script from FrameIntervals
                    if (ch == 0 && frameScriptMap.containsKey(fr)) {
                        displayName = frameScriptMap.get(fr);
                    } else {
                        displayName = resolveChannelCellName(dirFile, ch, chData);
                    }

                    data[ch][fr] = new ScoreCellData(
                            chData.castLib(),
                            chData.castMember(),
                            chData.spriteType(),
                            chData.ink(),
                            chData.posX(),
                            chData.posY(),
                            chData.width(),
                            chData.height(),
                            displayName
                    );
                }
            }
        }

        return data;
    }

    /**
     * Builds column names for the score table (frame numbers with optional labels).
     */
    public String[] buildColumnNames(DirectorFile dirFile) {
        ScoreChunk scoreChunk = dirFile.getScoreChunk();
        if (scoreChunk == null) {
            return new String[0];
        }

        int frameCount = scoreChunk.getFrameCount();
        String[] columnNames = new String[frameCount];

        // Get frame labels if available
        FrameLabelsChunk frameLabels = dirFile.getFrameLabelsChunk();
        Map<Integer, String> labelMap = new HashMap<>();
        if (frameLabels != null) {
            for (FrameLabelsChunk.FrameLabel label : frameLabels.labels()) {
                labelMap.put(label.frameNum(), label.label());
            }
        }

        for (int fr = 0; fr < frameCount; fr++) {
            String label = labelMap.get(fr + 1);
            if (label != null) {
                columnNames[fr] = (fr + 1) + " [" + label + "]";
            } else {
                columnNames[fr] = String.valueOf(fr + 1);
            }
        }

        return columnNames;
    }

    private Map<Integer, String> buildFrameScriptMap(DirectorFile dirFile, ScoreChunk scoreChunk, int frameCount) {
        Map<Integer, String> frameScriptMap = new HashMap<>();

        for (ScoreChunk.FrameInterval fi : scoreChunk.frameIntervals()) {
            var primary = fi.primary();
            var secondary = fi.secondary();
            if (secondary != null && primary.channelIndex() == 0) {
                String scriptName = resolveCastMemberByNumber(dirFile, secondary.castLib(), secondary.castMember());
                int startF = primary.startFrame() - 1;
                int endF = primary.endFrame() - 1;
                for (int f = startF; f <= endF && f >= 0 && f < frameCount; f++) {
                    frameScriptMap.put(f, scriptName);
                }
            }
        }

        return frameScriptMap;
    }

    private String resolveChannelCellName(DirectorFile dirFile, int channelIndex, ScoreChunk.ChannelData data) {
        switch (channelIndex) {
            case 0 -> {
                // Tempo channel - scripts come from FrameIntervals, not this data
                return "";
            }
            case 1 -> {
                // Palette channel - castMember is palette index
                int paletteId = data.castMember();
                return PaletteDescriptions.get(paletteId);
            }
            case 2 -> {
                // Transition channel - castMember is transition ID
                int transId = data.castMember();
                if (transId > 0) {
                    return "Trans #" + transId;
                }
                return "";
            }
            case 3, 4 -> {
                // Sound channels - these ARE cast member references
                return resolveCastMemberName(dirFile, data.castLib(), data.castMember());
            }
            case 5 -> {
                // Script channel - these ARE cast member references (frame scripts)
                return resolveCastMemberName(dirFile, data.castLib(), data.castMember());
            }
            default -> {
                // Regular sprite channels - cast member references
                return resolveCastMemberName(dirFile, data.castLib(), data.castMember());
            }
        }
    }

    private String resolveCastMemberName(DirectorFile dirFile, int castLib, int castMember) {
        CastMemberChunk member = dirFile.getCastMemberByIndex(castLib, castMember);

        if (member != null) {
            String name = member.name();
            if (name != null && !name.isEmpty()) {
                return "#" + member.id() + " " + name;
            }
            return "#" + member.id() + " " + member.memberType().getName();
        }

        return "Member #" + castMember;
    }

    private String resolveCastMemberByNumber(DirectorFile dirFile, int castLib, int memberNumber) {
        List<CastChunk> casts = dirFile.getCasts();
        if (casts.isEmpty()) {
            return "Member #" + memberNumber;
        }

        int index = memberNumber - 1;

        for (CastChunk cast : casts) {
            if (index >= 0 && index < cast.memberIds().size()) {
                int chunkId = cast.memberIds().get(index);
                if (chunkId != 0) {
                    Chunk chunk = dirFile.getChunk(chunkId);
                    if (chunk instanceof CastMemberChunk cm) {
                        String name = cm.name();
                        if (name != null && !name.isEmpty()) {
                            return "#" + cm.id() + " " + name;
                        }
                        return "#" + cm.id() + " " + cm.memberType().getName();
                    }
                }
            }
        }

        // Fallback
        Chunk chunk = dirFile.getChunk(memberNumber);
        if (chunk instanceof CastMemberChunk cm) {
            String name = cm.name();
            if (name != null && !name.isEmpty()) {
                return "#" + cm.id() + " " + name;
            }
            return "#" + cm.id() + " " + cm.memberType().getName();
        }

        return "Member #" + memberNumber;
    }
}

File: cpp/src/editor/score/ScoreDataBuilder.cpp
#include "libreshockwave/editor/score/ScoreDataBuilder.hpp"

#include <memory>
#include <utility>

#include "libreshockwave/cast/MemberType.hpp"
#include "libreshockwave/chunks/CastChunk.hpp"
#include "libreshockwave/chunks/CastMemberChunk.hpp"
#include "libreshockwave/chunks/FrameLabelsChunk.hpp"
#include "libreshockwave/editor/format/PaletteDescriptions.hpp"
#include "libreshockwave/id/Ids.hpp"

namespace libreshockwave::editor::score {
namespace {

std::string displayNameForMember(const chunks::CastMemberChunk& member) {
    if (!member.name().empty()) {
        return "#" + std::to_string(member.id().value()) + " " + member.name();
    }
    return "#" + std::to_string(member.id().value()) + " " + std::string(cast::name(member.memberType()));
}

} // namespace

ScoreDataBuilder::ScoreGrid ScoreDataBuilder::buildScoreData(DirectorFile& dirFile) const {
    const auto scoreChunk = dirFile.scoreChunk();
    if (scoreChunk == nullptr) {
        return {};
    }

    const int frameCount = scoreChunk->getFrameCount();
    const int channelCount = scoreChunk->getChannelCount();
    if (frameCount <= 0 || channelCount <= 0) {
        return {};
    }

    ScoreGrid data(static_cast<std::size_t>(channelCount));
    for (auto& row : data) {
        row.resize(static_cast<std::size_t>(frameCount));
    }

    const auto frameScriptMap = buildFrameScriptMap(dirFile, *scoreChunk, frameCount);
    for (const auto& entry : scoreChunk->frameData().frameChannelData) {
        const int frame = entry.frameIndex.value();
        const int channel = entry.channelIndex.value();
        const auto& channelData = entry.data;
        if (frame < 0 || frame >= frameCount || channel < 0 || channel >= channelCount || channelData.isEmpty()) {
            continue;
        }

        std::string displayName;
        if (channel == 0) {
            if (const auto script = frameScriptMap.find(frame); script != frameScriptMap.end()) {
                displayName = script->second;
            } else {
                displayName = resolveChannelCellName(dirFile, channel, channelData);
            }
        } else {
            displayName = resolveChannelCellName(dirFile, channel, channelData);
        }

        data[static_cast<std::size_t>(channel)][static_cast<std::size_t>(frame)] = model::ScoreCellData{
            channelData.castLib,
            channelData.castMember,
            channelData.spriteType,
            channelData.ink,
            channelData.posX,
            channelData.posY,
            channelData.width,
            channelData.height,
            std::move(displayName)
        };
    }

    return data;
}

std::vector<std::string> ScoreDataBuilder::buildColumnNames(const DirectorFile& dirFile) const {
    const auto scoreChunk = dirFile.scoreChunk();
    if (scoreChunk == nullptr) {
        return {};
    }

    const int frameCount = scoreChunk->getFrameCount();
    if (frameCount <= 0) {
        return {};
    }

    std::map<int, std::string> labelMap;
    if (const auto frameLabels = dirFile.frameLabelsChunk()) {
        for (const auto& label : frameLabels->labels()) {
            labelMap[label.frameNum.value()] = label.label;
        }
    }

    std::vector<std::string> columnNames;
    columnNames.reserve(static_cast<std::size_t>(frameCount));
    for (int frame = 0; frame < frameCount; ++frame) {
        const int frameNumber = frame + 1;
        if (const auto label = labelMap.find(frameNumber); label != labelMap.end()) {
            columnNames.push_back(std::to_string(frameNumber) + " [" + label->second + "]");
        } else {
            columnNames.push_back(std::to_string(frameNumber));
        }
    }
    return columnNames;
}

std::map<int, std::string> ScoreDataBuilder::buildFrameScriptMap(DirectorFile& dirFile,
                                                                 const chunks::ScoreChunk& scoreChunk,
                                                                 int frameCount) const {
    std::map<int, std::string> frameScriptMap;
    for (const auto& interval : scoreChunk.frameIntervals()) {
        if (!interval.secondary.has_value() || interval.primary.channelIndex != 0) {
            continue;
        }

        const auto scriptName = resolveCastMemberByNumber(
            dirFile,
            interval.secondary->castLib,
            interval.secondary->castMember);
        const int startFrame = interval.primary.startFrame - 1;
        const int endFrame = interval.primary.endFrame - 1;
        for (int frame = startFrame; frame <= endFrame; ++frame) {
            if (frame >= 0 && frame < frameCount) {
                frameScriptMap[frame] = scriptName;
            }
        }
    }
    return frameScriptMap;
}

std::string ScoreDataBuilder::resolveChannelCellName(DirectorFile& dirFile,
                                                     int channelIndex,
                                                     const chunks::ScoreChunk::ChannelData& data) const {
    switch (channelIndex) {
        case 0:
            return "";
        case 1:
            return format::PaletteDescriptions::get(data.castMember);
        case 2:
            if (data.castMember > 0) {
                return "Trans #" + std::to_string(data.castMember);
            }
            return "";
        case 3:
        case 4:
        case 5:
        default:
            return resolveCastMemberName(dirFile, data.castLib, data.castMember);
    }
}

std::string ScoreDataBuilder::resolveCastMemberName(DirectorFile& dirFile, int castLib, int castMember) const {
    const auto member = dirFile.getCastMemberByIndex(castLib, castMember);
    if (member != nullptr) {
        return displayNameForMember(*member);
    }
    return "Member #" + std::to_string(castMember);
}

std::string ScoreDataBuilder::resolveCastMemberByNumber(DirectorFile& dirFile, int castLib, int memberNumber) const {
    (void)castLib;
    const int index = memberNumber - 1;
    for (const auto& cast : dirFile.casts()) {
        if (cast == nullptr || index < 0 || index >= static_cast<int>(cast->memberIds().size())) {
            continue;
        }
        const int chunkId = cast->memberIds()[static_cast<std::size_t>(index)];
        if (chunkId == 0) {
            continue;
        }
        const auto member = std::dynamic_pointer_cast<chunks::CastMemberChunk>(dirFile.getChunk(id::ChunkId(chunkId)));
        if (member != nullptr) {
            return displayNameForMember(*member);
        }
    }

    const auto fallback = std::dynamic_pointer_cast<chunks::CastMemberChunk>(dirFile.getChunk(id::ChunkId(memberNumber)));
    if (fallback != nullptr) {
        return displayNameForMember(*fallback);
    }
    return "Member #" + std::to_string(memberNumber);
}

} // namespace libreshockwave::editor::score
